### PR TITLE
Add configurable rez spell list

### PR DESCRIPTION
**Pull request by acastle**
Remove the hard coded list of rez spells in favor of a list which can be configured from the character ini.

#72

### CODE REVIEW

File: E3Next/Settings/CharacterSettings.cs [hunk 1]
```diff
@@ -133,6 +133,7 @@ public class CharacterSettings : BaseSettings, IBaseSettings
         public List<Spell> HealXTarget = new List<Spell>();
         public List<Spell> HealPets = new List<Spell>();
         public List<Spell> HealOverTime = new List<Spell>();
+        public List<Spell> HealResurrection = new List<Spell>();
```
Comment (jonnyrocket0) [inline E3Next/Settings/CharacterSettings.cs:136]: Change this to something more descriptive like `RezSpells` or `ResurrectionSpells`

File: E3Next/Settings/CharacterSettings.cs [hunk 2]
```diff
@@ -330,6 +331,7 @@ public void LoadData()
             LoadKeyData("Heals", "Heal Over Time Spell", ParsedData, HealOverTime);
             LoadKeyData("Heals", "Group Heal", ParsedData, HealGroup);
             LoadKeyData("Heals", "Pet Heal", ParsedData, HealPets);
+            LoadKeyData("Heals", "Resurrection", ParsedData, HealResurrection);
```
Comment (jonnyrocket0) [inline E3Next/Settings/CharacterSettings.cs:334]: Change the key name to something more descriptive like `RezSpells` or `ResurrectionSpells`

File: E3Next/Settings/CharacterSettings.cs [hunk 3]
```diff
@@ -538,6 +540,7 @@ public IniData CreateSettings(string fileName)
                 section.Keys.AddKey("Who to HoT", "");
                 section.Keys.AddKey("Pet Owner", "");
                 section.Keys.AddKey("Auto Cast Necro Heal Orbs (On/Off)", "On");
+                section.Keys.AddKey("Resurrection", "");
```
Comment (jonnyrocket0) [inline E3Next/Settings/CharacterSettings.cs:543]: Default this to the list you pulled out of the `Rez` class

File: E3Next/Processors/Rez.cs [hunk 4]
```diff
@@ -401,11 +392,9 @@ private static void MultiRez(RezType rezType = RezType.AE)
 
                     //assume consent was given
                     MQ.Cmd("/corpse");
-
-               
-                    foreach (var spell in _currentRezSpells)
+                    
+                    foreach (var spell in E3.CharacterSettings.HealResurrection)
```
Comment (jonnyrocket0) [inline E3Next/Processors/Rez.cs:396]: In this loop, if it's a rez token, validate we have one

Comment (acastle) [reply E3Next/Processors/Rez.cs:396]: I believe this is already handled via a combination of the constructor for `Data.Spell` which will populate the initial item data from MQ queries on first initialization and `Casting.CheckReady` which returns true only when `Me.ItemReady[<name>]` evals to true.

https://github.com/RekkasGit/E3Next/blob/fdc16a033bc6f067692b6078d8c79102a5b35dc2/E3Next/Processors/Casting.cs#L894

Comment (jonnyrocket0) [reply E3Next/Processors/Rez.cs:396]: Ah I think you're right. Good call out.

Comment (RekkasGit) [reply E3Next/Processors/Rez.cs:396]: I need to look at this closer, but I believe you will DC from the Wait4Rez tell being spammed, as yes the checkready may fail, but you need to kickout long before you reach that point.  This was the reason for the Refresh Spell method.

File: E3Next/Settings/CharacterSettings.cs (hunk 3 above)
Comment (acastle) [reply E3Next/Settings/CharacterSettings.cs:543]: Working on this now. Seems to be a bug with merging sections of ini files with multiple values that will need to be fixed to render out the multiple options.

File: E3Next/Processors/Rez.cs (hunk 4 above)
Comment (acastle) [reply E3Next/Processors/Rez.cs:396]: You might be right there, I'll look into it there is a clean way to throttle the attempts

Comment (RekkasGit) [reply E3Next/Processors/Rez.cs:396]: Less about throttle, more of a check if its in the spell book, if not, check if its an item.. if not (we are out of tokens), skip it.

**Comment by acastle** [comment]
Sorry, this PR ended up getting a bit bigger than I was hoping but at least there are some unit tests 😅

Ended up having to tackle a couple of bug fixes in the process, specifically around writing ini files when there are multiple values (current behavior only takes the last value). The other is a stack overflow on the spells ToString method due to the Before/After spell data props. I am actually surprised that no one tripped over the ToString one to this point, for me at least it causes both the VS2022 and Rider debuggers to crash if you ever put a breakpoint in Spell.cs. I must just be _lucky_.

Anywho, should be ready for another set of 👀 whenever yall have a moment.

**Comment by RekkasGit** [comment]
Will take a look at this tomorrow, but just off the top of my head, saving multiple settings was already done in Character Settings under SaveData(). You simply add to the section, and the writer figures out that you want to write it all out to disk. (a little hacky but wanted to keep very minimal changes to the IniFileParser)
I am a little nervous having somone go into the ini area as its very easy to break things, but again will take a look at it tomorrow.

**Comment by RekkasGit** [comment]
Sorry for the delay, I will get to it this week.

**Comment by RekkasGit** [comment]
I've taken parts of this PR and a few edits as well and created another PR and its currently in master. 
Closing this PR for now. Ty for your work on this:)

**Pull request closed without merging**